source/common/quic/envoy_quic_session_cache.cc

```diff
@@ -0,0 +1,168 @@
+    return;
+  }
+  Entry& entry = it->second;
+  if (params == *entry.params &&
+      doApplicationStatesMatch(application_state, entry.application_state.get())) {
+    // The states are both the same, so we only need to insert the session.
+    entry.pushSession(std::move(session));
+    return;
+  }
+  // States are different, replace the entry.
+  cache_.erase(it);
+  createAndInsertEntry(server_id, std::move(session), params, application_state);
+}
+
+std::unique_ptr<quic::QuicResumptionState>
+EnvoyQuicSessionCache::Lookup(const quic::QuicServerId& server_id, const SSL_CTX* /*ctx*/) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return nullptr;
+  }
+  Entry& entry = it->second;
+  const SystemTime system_time = time_source_.systemTime();
+  if (!isSessionValid(entry.peekSession(), system_time)) {
+    cache_.erase(it);
+    return nullptr;
+  }
+  auto state = std::make_unique<quic::QuicResumptionState>();
+  state->tls_session = entry.popSession();
+  if (entry.params != nullptr) {
+    state->transport_params = std::make_unique<quic::TransportParameters>(*entry.params);
+  }
+  if (entry.application_state != nullptr) {
+    state->application_state = std::make_unique<quic::ApplicationState>(*entry.application_state);
+  }
+  return state;
+}
+
+void EnvoyQuicSessionCache::ClearEarlyData(const quic::QuicServerId& server_id) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return;
+  }
+  for (bssl::UniquePtr<SSL_SESSION>& session : it->second.sessions) {
+    if (session != nullptr) {
+      session.reset(SSL_SESSION_copy_without_early_data(session.get()));
+    }
+  }
+}
+
+void EnvoyQuicSessionCache::prune() {
+  quic::QuicServerId oldest_id;
+  uint64_t oldest_expiration = std::numeric_limits<uint64_t>::max();
+  const SystemTime system_time = time_source_.systemTime();
+  auto it = cache_.begin();
+  while (it != cache_.end()) {
+    Entry& entry = it->second;
+    SSL_SESSION* session = entry.peekSession();
+    if (!isSessionValid(session, system_time)) {
+      it = cache_.erase(it);
+    } else {
+      if (cache_.size() >= kMaxSessionCacheEntries) {
+        // Only save oldest if we are at the size limit.
+        const uint64_t session_expiration =
+            SSL_SESSION_get_time(session) + SSL_SESSION_get_timeout(session);
+        if (session_expiration < oldest_expiration) {
+          oldest_expiration = session_expiration;
+          oldest_id = it->first;
+        }
+      }
+      ++it;
+    }
+  }
+  if (cache_.size() >= kMaxSessionCacheEntries) {
+    cache_.erase(oldest_id);
+  }
+}
+
+void EnvoyQuicSessionCache::createAndInsertEntry(const quic::QuicServerId& server_id,
+                                                 bssl::UniquePtr<SSL_SESSION> session,
+                                                 const quic::TransportParameters& params,
+                                                 const quic::ApplicationState* application_state) {
```

Comment:
It's a bit confusing about the life time of application_state. Can you comment that it's now owned?

Reply:
The input parameter owned. The fact that it's a const pointer makes the ownership model clear to me. Am I missing something?

Reply:
> sorry, typo "not" for "now".

I don't think we need to comment that a const pointer is not owned, as that's very common usage.

> If we create ApplicationState instance for each entry here, why does doApplicationStatesMatch() compare pointers?

doApplicationStatesMatch() dereferences the pointers and compares the values. It looks at pointers because they can be nullptr.

> And Lookup() also create ApplicationState for return value. Probably QUICHE somehow requires duplication this object.

I'm not sure what you're asking? QUICHE passes us a const pointer and we need to make a copy that we own.